src/libpmem/pmem.c

```diff
@@ -1282,3 +1589,70 @@ pmem_init(void)
 			"QueryVirtualMemoryInformation");
 #endif
 }
+
+
+/*
+ * range_deep_flush -- (internal) perform deep flush of given address range
+ */
+static int
+range_deep_flush(uintptr_t addr, size_t len)
+{
+
+	while (len != 0) {
+		const struct map_tracker *mt = map_range_find(addr, len);
+
+		if (mt == NULL) /* no more overlapping track regions */
+			return pmem_msync((void *)addr, len);
+
+		if (mt->base_addr > addr) {
+			size_t curr_len = mt->base_addr - addr;
+			if (curr_len > len)
+				curr_len = len;
+			if (pmem_msync((void *)addr, curr_len) != 0)
+				return -1;
+			if ((len -= curr_len) == 0)
+				return 0;
+			addr = mt->base_addr;
+		}
+
+		if (ddax_deep_flush_write(mt->region_id) < 0) {
+			ERR("!cannot perform final write to deep_flush");
+			return -1;
+		}
+
+		if (mt->end_addr >= addr + len)
+			return 0;
+
+		len -= mt->end_addr - addr;
+		addr = mt->end_addr;
+	}
+	return 0;
+}
+
+
+/*
+ * pmem_deep_flush -- perform deep flush on a memory range
+ *
+ * It merely acts as wrapper around an msync call in most cases, the only
+ * exception is the case of an mmap'ed DAX device on Linux.
+ */
+int
+pmem_deep_flush(const void *addr, size_t len)
+{
+	LOG(3, "addr %p len %zu", addr, len);
+
+	if (len == 0)
+		return 0;
+
+	if (os_rwlock_rdlock(&Mmap_list_lock)) {
```

Comment:
Is there are reason why you need to grab this particular lock here? The "map_" something function called deeper in the stack should grab this lock.